ERCS/erc-7546.md

```diff
@@ -0,0 +1,187 @@
+
+#### 1.1 Storage & Events
+The Dictionary MUST maintain a mapping of function selectors to implementation addresses.
+
+Changes to this mapping SHOULD be communicated through an event (or log).
+```solidity
+event ImplementationUpgraded(bytes4 indexed functionSelector, address indexed implementation);
+```
+
+#### 1.2 Functions
+The Dictionary SHOULD provide functions to retrieve and set implementations.
+
+##### 1.2.1 `getImplementation`
+```solidity
+function getImplementation(bytes4 functionSelector) external view returns (address);
+```
+
+##### 1.2.2 `setImplementation`
+```solidity
+function setImplementation(bytes4 functionSelector, address implementation) external;
+```
+The implementation contract MUST implement the function selector to be registered as a key in the mapping (section 3.1). Therefore, it is RECOMMENDED to verify this during registration.
+
+### 2. Proxy Contract
+This contract checks the Dictionary for the associated implementation address based on its function selector and delegatecalls to it.
+
+#### 2.1 Storage & Events
+The Proxy MUST reference the Dictionary address while ensuring a minimal chance of storage clashes. This is because the Dictionary's registered implementations will use this shared storage.
+
+Changes to the Dictionary address SHOULD be noticed:
+```solidity
+event DictionaryUpgraded(address indexed dictionary);
+```
+
+While it is possible to store the Dictionary address in the code area (e.g., using Solidity's immutable or constant), it SHOULD be designed with caution, considering the possibility that if the Dictionary's admin is not the same as the Proxy's admin, the ability to manipulate the implementation could be permanently lost.
+
+#### 2.2 Functions
+All calls to the Proxy MUST be forwarded to the implementation address registered in the Dictionary through the fallback function. Furthermore, as there is a possibility of collisions with function selectors registered in the Dictionary, Proxy SHOULD NOT have external functions.
+
+##### 2.2.1 `constructor`
+It is RECOMMENDED to initialize contracts, other than setting up the Dictionary address, through the Dictionary as well.
+
+##### 2.2.2 `fallback`
+The fallback function MUST have the functionality of calling the implementation address using calldata and delegatecall, and processing the return value. The implementation address to which delegatecall MUST be the returned value queried to the Dictionary (stored in the slot specified in section 2.1) using the `getImplementation(bytes4 functionSelector)` function defined in section 1.2.1.
+
+
+### 3. Interfaces
+The implementation contract MUST have the same function selector registered in the Dictionary. If not, the Proxy's delegatecall will fail.
+
+For transparency, it's RECOMMENDED for the Proxy to support interface queries from both on-chain and off-chain entities.
```

Comment:
What is a query, and how would a contract implement it differently for on-/off-chain entities?

Reply:
Due to the lack of clarity in the content, we have made significant revisions to this section.